Compile to WASM (WebAssembly) Module with Emscripten

Issue overview
--------------
I was trying to compile to a WASM (WebAssembly) Module with Emscripten.

### Details
These are steps I followed:

1. Installed Emscripton on my system. emcc --check produces the output
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.37-git
shared:INFO: (Emscripten: Running sanity checks)

2. Cloned the EnergyPlus from GitHub git clone https://github.com/NREL/EnergyPlus
3. `cd EnergyPlus`
4. `mkdir build`
5. `cd build`
6. `emcmake cmake ..` This gave the error
```
-- Using X11 for window creation
CMake Error at /Applications/CMake.app/Contents/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find X11 (missing: X11_X11_LIB)
Call Stack (most recent call first):
  /Applications/CMake.app/Contents/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /Applications/CMake.app/Contents/share/cmake-3.26/Modules/FindX11.cmake:481 (find_package_handle_standard_args)
  third_party/penumbra/vendor/glfw/CMakeLists.txt:196 (find_package)
```
so I instead tried `cmake .. -DGLFW_USE_OSMESA=1 -DBUILD_STATIC_ENERGYPLUS_API=1` and this now give me the error
```
CMake Error in src/EnergyPlus/CMakeLists.txt:
  export called with target "energyplusapi" which requires target
  "energypluslib" that is not in any export set.


CMake Error in src/EnergyPlus/CMakeLists.txt:
  export called with target "energyplusapi" which requires target
  "project_options" that is not in any export set.


CMake Error in src/EnergyPlus/CMakeLists.txt:
  export called with target "energyplusapi" which requires target
  "project_fp_options" that is not in any export set.


CMake Error in src/EnergyPlus/CMakeLists.txt:
  export called with target "energyplusapi" which requires target
  "project_warnings" that is not in any export set.
```

### Build Machine
- I am using a MacBook Pro M1 and Emscripten version
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.37-git
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

@jmarrec Many thanks! I tried these steps and here are the logs from the `ninja` while building in the last step for `EnergyPlus-build-static`

```bash
➜  EnergyPlus-build-static ninja                             
[0/1] Re-running CMake...
CMake Warning (dev) at CMakeLists.txt:10 (project):
  cmake_minimum_required() should be called prior to this top-level project()
  call.  Please see the cmake-commands(7) manual for usage documentation of
  both commands.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Adding -fcolor-diagnostics
-- JsonCpp Version: 1.9.3
-- Module support is disabled.
-- Version: 8.0.1
-- Build type: Release
-- CXX_STANDARD: 17
-- Required features: cxx_variadic_templates
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
-- Using OSMesa for headless context creation
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND CXX) 
-- The target `docs` was added, but not to the default build rule. It will be triggered by `make docs` or `make package`
-- Configuring done (9.4s)
-- Generating done (0.1s)
-- Build files have been written to: /Users/sid/EnergyPlus-build-static
[2/298] Building CXX object src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o
FAILED: src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o 
/opt/homebrew/Cellar/emscripten/3.1.37/libexec/em++ -DENERGYPLUS_STATIC_API -DEP_cache_GlycolSpecificHeat -DEP_psych_errors -DJSON_USE_IMPLICIT_CONVERSIONS=OFF -DOBJEXXFCL_ALIGN=64 -DXML_STATIC -I/Users/sid/EnergyPlus/nlohmann -I/Users/sid/EnergyPlus/third_party/re2 -I/Users/sid/EnergyPlus/src -I/Users/sid/EnergyPlus/third_party -I/Users/sid/EnergyPlus/third_party/fmt-8.0.1/include -I/Users/sid/EnergyPlus/third_party/btwxt/src -I/Users/sid/EnergyPlus/third_party/doj -I/Users/sid/EnergyPlus/third_party/nlohmann -I/Users/sid/EnergyPlus/third_party/fast_float -I/Users/sid/EnergyPlus/third_party/valijson -I/Users/sid/EnergyPlus/third_party/ObjexxFCL/src -I/Users/sid/EnergyPlus/third_party/CLI -I/Users/sid/EnergyPlus/third_party/eigen -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Chromogenics/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Common/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Gases/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/MultiLayerOptics/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/SingleLayerOptics/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/SpectralAveraging/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Tarcog/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Viewer/include -I/Users/sid/EnergyPlus/third_party/ssc -I/Users/sid/EnergyPlus/third_party/penumbra/include -I/Users/sid/EnergyPlus/third_party/kiva/src -I/Users/sid/EnergyPlus-build-static/third_party/kiva/src/libkiva -I/Users/sid/EnergyPlus/third_party/cpgfunctionEP -I/Users/sid/EnergyPlus/third_party/cpgfunctionEP/include -isystem /Users/sid/EnergyPlus/third_party/SQLite -isystem /Users/sid/EnergyPlus/third_party/Expat -isystem /Users/sid/EnergyPlus/third_party/Expat/lib -isystem /Users/sid/EnergyPlus/third_party/kiva/vendor/boost-1.61.0 -O3 -DNDEBUG -std=gnu++17 -fPIC -pipe -fno-stack-protector -fcolor-diagnostics -Wpedantic -Wall -Wextra -Wno-unknown-pragmas -Wno-attributes -Wno-delete-non-virtual-dtor -Wno-missing-braces -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wno-vexing-parse -Wno-invalid-source-encoding -MD -MT src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o -MF src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o.d -o src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o -c /Users/sid/EnergyPlus-build-static/src/EnergyPlus/EmbeddedEpJSONSchema.cc
/Users/sid/EnergyPlus-build-static/src/EnergyPlus/EmbeddedEpJSONSchema.cc:1:1: error: unknown type name 'Will'
Will be filled in later by GenerateEmbeddedEpJSONSchema
^
/Users/sid/EnergyPlus-build-static/src/EnergyPlus/EmbeddedEpJSONSchema.cc:1:8: error: expected ';' after top level declarator
Will be filled in later by GenerateEmbeddedEpJSONSchema
       ^
       ;
2 errors generated.
em++: error: '/opt/homebrew/Cellar/emscripten/3.1.37/libexec/llvm/bin/clang++ -target wasm32-unknown-emscripten -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -DEMSCRIPTEN --sysroot=/opt/homebrew/Cellar/emscripten/3.1.37/libexec/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -DENERGYPLUS_STATIC_API -DEP_cache_GlycolSpecificHeat -DEP_psych_errors -DJSON_USE_IMPLICIT_CONVERSIONS=OFF -DOBJEXXFCL_ALIGN=64 -DXML_STATIC -I/Users/sid/EnergyPlus/nlohmann -I/Users/sid/EnergyPlus/third_party/re2 -I/Users/sid/EnergyPlus/src -I/Users/sid/EnergyPlus/third_party -I/Users/sid/EnergyPlus/third_party/fmt-8.0.1/include -I/Users/sid/EnergyPlus/third_party/btwxt/src -I/Users/sid/EnergyPlus/third_party/doj -I/Users/sid/EnergyPlus/third_party/nlohmann -I/Users/sid/EnergyPlus/third_party/fast_float -I/Users/sid/EnergyPlus/third_party/valijson -I/Users/sid/EnergyPlus/third_party/ObjexxFCL/src -I/Users/sid/EnergyPlus/third_party/CLI -I/Users/sid/EnergyPlus/third_party/eigen -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Chromogenics/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Common/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Gases/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/MultiLayerOptics/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/SingleLayerOptics/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/SpectralAveraging/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Tarcog/include -I/Users/sid/EnergyPlus/third_party/Windows-CalcEngine/src/Viewer/include -I/Users/sid/EnergyPlus/third_party/ssc -I/Users/sid/EnergyPlus/third_party/penumbra/include -I/Users/sid/EnergyPlus/third_party/kiva/src -I/Users/sid/EnergyPlus-build-static/third_party/kiva/src/libkiva -I/Users/sid/EnergyPlus/third_party/cpgfunctionEP -I/Users/sid/EnergyPlus/third_party/cpgfunctionEP/include -isystem /Users/sid/EnergyPlus/third_party/SQLite -isystem /Users/sid/EnergyPlus/third_party/Expat -isystem /Users/sid/EnergyPlus/third_party/Expat/lib -isystem /Users/sid/EnergyPlus/third_party/kiva/vendor/boost-1.61.0 -O3 -DNDEBUG -std=gnu++17 -fPIC -pipe -fno-stack-protector -fcolor-diagnostics -Wpedantic -Wall -Wextra -Wno-unknown-pragmas -Wno-attributes -Wno-delete-non-virtual-dtor -Wno-missing-braces -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wno-vexing-parse -Wno-invalid-source-encoding -MD -MT src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o -MF src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o.d -c /Users/sid/EnergyPlus-build-static/src/EnergyPlus/EmbeddedEpJSONSchema.cc -o src/EnergyPlus/CMakeFiles/energyplusparser.dir/EmbeddedEpJSONSchema.cc.o' failed (returned 1)
[11/298] Building CXX object src/EnergyPlus/CMakeFiles/energypluslib.dir/AirLoopHVACDOAS.cc.o
ninja: build stopped: subcommand failed.
```

--- a/src/EnergyPlus/CMakeLists.txt
+++ b/src/EnergyPlus/CMakeLists.txt
@@ -875,45 +875,47 @@ add_custom_command(
     -DPYTHON_API_VERSION_MAJOR="${PYTHON_API_VERSION_MAJOR}" -DPYTHON_API_VERSION_MINOR="${PYTHON_API_VERSION_MINOR}" -P
     "${PROJECT_SOURCE_DIR}/cmake/PythonSetupAPIinBuild.cmake")
 
-install(
-  TARGETS energyplusapi energyplus
-  EXPORT energyplus
-  RUNTIME DESTINATION ./
-  LIBRARY DESTINATION ./
-  ARCHIVE DESTINATION ./
-  INCLUDES DESTINATION include
-                       include/EnergyPlus/api
-)
+if(NOT BUILD_STATIC_ENERGYPLUS_API)
+  install(
+    TARGETS energyplusapi energyplus
+    EXPORT energyplus
+    RUNTIME DESTINATION ./
+    LIBRARY DESTINATION ./
+    ARCHIVE DESTINATION ./
+    INCLUDES DESTINATION include
+                         include/EnergyPlus/api
+  )
 
-include(CMakePackageConfigHelpers)
-write_basic_package_version_file(
-  "${PROJECT_BINARY_DIR}/energyplusConfigVersion.cmake"
-  VERSION ${ENERGYPLUS_VERSION}
-  COMPATIBILITY AnyNewerVersion
-)
-
-export(
-  EXPORT
-  energyplus
-       NAMESPACE "energyplus::"
-       FILE "${PROJECT_BINARY_DIR}/energyplusConfig.cmake"
-)
-
-install(
-  EXPORT energyplus
-        DESTINATION lib/cmake/energyplus
-        NAMESPACE "energyplus::"
-        COMPONENT "CPPAPI"
-  FILE energyplusConfig.cmake
-)
-
-install(
-  FILES
+  include(CMakePackageConfigHelpers)
+  write_basic_package_version_file(
     "${PROJECT_BINARY_DIR}/energyplusConfigVersion.cmake"
-  DESTINATION
-    lib/cmake/energyplus/
-  COMPONENT "CPPAPI"
-)
+    VERSION ${ENERGYPLUS_VERSION}
+    COMPATIBILITY AnyNewerVersion
+  )
+
+  export(
+    EXPORT
+    energyplus
+         NAMESPACE "energyplus::"
+         FILE "${PROJECT_BINARY_DIR}/energyplusConfig.cmake"
+  )
+
+  install(
+    EXPORT energyplus
+          DESTINATION lib/cmake/energyplus
+          NAMESPACE "energyplus::"
+          COMPONENT "CPPAPI"
+    FILE energyplusConfig.cmake
+  )
+
+  install(
+    FILES
+      "${PROJECT_BINARY_DIR}/energyplusConfigVersion.cmake"
+    DESTINATION
+      lib/cmake/energyplus/
+    COMPONENT "CPPAPI"
+  )
+endif()
 
 if(LINK_WITH_PYTHON)
   # link the executable to the dll and bring it in

